File: atcoder/dp/b-frog-2.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <algorithm>
const int INF = 1e9 + 5;
using namespace std;
int main() {
    int n, k;
    cin >> n >> k;
    vector<int> h(n);
    for (auto &x : h) {
        cin >> x;
    }

    vector<int> dp(n, INF);
    dp[0] = 0;
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j <= i + k; j++) {
            if (j < n) {
                dp[j] = min(dp[j], dp[i] + abs(h[i] - h[j]));
            }
        } 
    }
    cout << dp[n - 1] << endl;

    return 0;
}

File: atcoder/dp/B-Frog-2.py
N,K = map(int, input().split())
h = list(map(int, input().split()))
INF = 10**9

dp = [INF]*(N)
dp[0] = 0

for i in range (N):
    for j in range(1, K+1):
        if i+j < N:
            dp[i+j] = min(abs(h[i] - h[i+j]) + dp[i], dp[i+j])
print(dp[N-1])
